File: src/node/src/clients/halo-infinite-client.ts
import { ClientBase } from './base/client-base';
import type { HaloInfiniteClientOptions } from '../models/common/client-options';

// Module imports
import { AcademyModule } from '../modules/halo-infinite/academy.module';
import { BanProcessorModule } from '../modules/halo-infinite/ban-processor.module';
import { ConfigurationModule } from '../modules/halo-infinite/configuration.module';
import { EconomyModule } from '../modules/halo-infinite/economy.module';
import { GameCmsModule } from '../modules/halo-infinite/game-cms.module';
import { LobbyModule } from '../modules/halo-infinite/lobby.module';
import { SettingsModule } from '../modules/halo-infinite/settings.module';
import { SkillModule } from '../modules/halo-infinite/skill.module';
import { StatsModule } from '../modules/halo-infinite/stats.module';
import { TextModerationModule } from '../modules/halo-infinite/text-moderation.module';
import { UgcModule } from '../modules/halo-infinite/ugc.module';
import { UgcDiscoveryModule } from '../modules/halo-infinite/ugc-discovery.module';

/**
 * Main client for interacting with Halo Infinite APIs.
 *
 * Provides access to all API functionality through domain-specific modules.
 * Modules are lazily initialized on first access to minimize memory usage.
 *
 * @example
 * ```typescript
 * import { HaloInfiniteClient, MatchType, LifecycleMode, isSuccess } from '@dendev/grunt';
 *
 * // Create client with authentication
 * const client = new HaloInfiniteClient({
 *   spartanToken: 'your-spartan-token',
 *   xuid: '2533274855333605',
 *   clearanceToken: 'flight-clearance-id',
 * });
 *
 * // Get match history
 * const history = await client.stats.getMatchHistory('2533274855333605', 0, 25, MatchType.All);
 * if (isSuccess(history)) {
 *   console.log(`Found ${history.result.resultCount} matches`);
 * }
 *
 * // Get player inventory
 * const inventory = await client.economy.getInventoryItems('2533274855333605');
 *
 * // Get CSR for a playlist
 * const csr = await client.skill.getPlaylistCsr('playlist-id', ['2533274855333605']);
 *
 * // Search for UGC maps
 * const maps = await client.ugcDiscovery.search({
 *   assetKinds: [AssetKind.Map],
 *   term: 'blood gulch',
 * });
 * ```
 */
export class HaloInfiniteClient extends ClientBase {
  // Lazy-loaded module instances
  private _academy?: AcademyModule;
  private _banProcessor?: BanProcessorModule;
  private _configuration?: ConfigurationModule;
  private _economy?: EconomyModule;
  private _gameCms?: GameCmsModule;
  private _lobby?: LobbyModule;
  private _settings?: SettingsModule;
  private _skill?: SkillModule;
  private _stats?: StatsModule;
  private _textModeration?: TextModerationModule;
  private _ugc?: UgcModule;
  private _ugcDiscovery?: UgcDiscoveryModule;

  /**
   * Creates a new HaloInfiniteClient instance.
   *
   * @param options - Client configuration options
   *
   * @example
   * ```typescript
   * const client = new HaloInfiniteClient({
   *   spartanToken: 'your-spartan-token',
   *   xuid: '2533274855333605',
   *   includeRawResponses: true, // Enable for debugging
   * });
   * ```
   */
  constructor(options: HaloInfiniteClientOptions) {
    super({
      fetchFn: options.fetchFn,
      cacheTtlMs: options.cacheTtlMs,
      maxRetries: options.maxRetries,
    });

    this.spartanToken = options.spartanToken;
    this.xuid = options.xuid ?? '';
    this.clearanceToken = options.clearanceToken ?? '';
    this.includeRawResponses = options.includeRawResponses ?? false;
    this.userAgent = options.userAgent ?? '';
  }

  /**
   * Academy module for bot customization and drill-related APIs.
   *
   * Provides access to:
   * - Bot customization options
   * - Academy content manifest
   * - Star/scoring definitions for drills
   */
  get academy(): AcademyModule {
    return (this._academy ??= new AcademyModule(this));
  }

  /**
   * Ban Processor module for ban-related APIs.
   *
   * Provides access to:
   * - Ban summary queries for players
   */
  get banProcessor(): BanProcessorModule {
    return (this._banProcessor ??= new BanProcessorModule(this));
  }

  /**
   * Configuration module for endpoint discovery APIs.
   *
   * Provides access to:
   * - API settings and endpoint configuration
   */
  get configuration(): ConfigurationModule {
    return (this._configuration ??= new ConfigurationModule(this));
  }

  /**
   * Economy module for player customization, stores, and inventory APIs.
   *
   * Provides access to:
   * - Player inventory and currency balances
   * - Customization (armor, weapons, vehicles, AI)
   * - In-game stores and offerings
   * - Active boosts and rewards
   * - Operation/battle pass progress
   */
  get economy(): EconomyModule {
    return (this._economy ??= new EconomyModule(this));
  }

  /**
   * Game CMS module for static content and definitions.
   *
   * Provides access to:
   * - Item definitions
   * - Challenge definitions
   * - Season and career metadata
   * - Medal information
   * - News and guides
   */
  get gameCms(): GameCmsModule {
    return (this._gameCms ??= new GameCmsModule(this));
  }

  /**
   * Lobby module for multiplayer lobby and presence APIs.
   *
   * Provides access to:
   * - QoS servers
   * - Player presence in lobbies
   * - Join handles and lobby joining
   */
  get lobby(): LobbyModule {
    return (this._lobby ??= new LobbyModule(this));
  }

  /**
   * Settings module for clearance and flight configuration APIs.
   *
   * Provides access to:
   * - Clearance levels and feature flags
   */
  get settings(): SettingsModule {
    return (this._settings ??= new SettingsModule(this));
  }

  /**
   * Skill module for CSR (Competitive Skill Rank) APIs.
   *
   * Provides access to:
   * - Match skill results (CSR changes after a match)
   * - Playlist CSR for players
   */
  get skill(): SkillModule {
    return (this._skill ??= new SkillModule(this));
  }

  /**
   * Stats module for match history and service record APIs.
   *
   * Provides access to:
   * - Match history for players
   * - Individual match statistics
   * - Player service records (career stats)
   * - Challenge decks and progression
   */
  get stats(): StatsModule {
    return (this._stats ??= new StatsModule(this));
  }

  /**
   * Text Moderation module for moderation-related APIs.
   *
   * Provides access to:
   * - Moderation keys for text validation
   */
  get textModeration(): TextModerationModule {
    return (this._textModeration ??= new TextModerationModule(this));
  }

  /**
   * UGC (User Generated Content) module for authoring operations.
   *
   * Provides access to:
   * - Creating, editing, and deleting user content
   * - Managing asset permissions
   * - Rating and favoriting assets
   * - Publishing and unpublishing assets
   */
  get ugc(): UgcModule {
    return (this._ugc ??= new UgcModule(this));
  }

  /**
   * UGC Discovery module for searching and browsing user content.
   *
   * Provides access to:
   * - Searching for maps, game variants, and other content
   * - Browsing featured and popular content
   * - Getting recommended content
   */
  get ugcDiscovery(): UgcDiscoveryModule {
    return (this._ugcDiscovery ??= new UgcDiscoveryModule(this));
  }
}

File: src/dotnet/Den.Dev.Grunt/Den.Dev.Grunt/Core/HaloInfiniteClient.cs
// <copyright file="HaloInfiniteClient.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Den.Dev.Grunt is managed by Halo Studios and Microsoft. This wrapper is not endorsed by Halo Studios or Microsoft.
// </copyright>

using System.Net.Http;
using Den.Dev.Grunt.Core.Foundation;
using Den.Dev.Grunt.Core.Modules;
using Den.Dev.Grunt.Core.Modules.HaloInfinite;

namespace Den.Dev.Grunt.Core
{
    /// <summary>
    /// Client for interacting with Halo Infinite APIs through domain-specific modules.
    /// </summary>
    /// <remarks>
    /// This client provides organized access to Halo Infinite APIs through specialized modules.
    /// Each module handles a specific domain of functionality (Economy, GameCms, Stats, etc.).
    ///
    /// <example>
    /// Example usage:
    /// <code>
    /// var client = new HaloInfiniteClient(spartanToken, xuid, clearanceToken);
    ///
    /// // Access economy APIs
    /// var inventory = await client.Economy.GetInventoryItemsAsync(player);
    ///
    /// // Access stats APIs
    /// var matchHistory = await client.Stats.GetMatchHistoryAsync(player, 0, 25, MatchType.All);
    ///
    /// // Access GameCMS APIs
    /// var medals = await client.GameCms.GetMedalMetadataAsync();
    /// </code>
    /// </example>
    /// </remarks>
    public sealed class HaloInfiniteClient : ClientBase, IHaloInfiniteClient
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="HaloInfiniteClient"/> class.
        /// </summary>
        /// <param name="spartanToken">The Spartan token used for authentication.</param>
        /// <param name="xuid">Optional Xbox User ID.</param>
        /// <param name="clearanceToken">Optional clearance token for flighted content.</param>
        /// <param name="includeRawResponses">Whether to include raw API responses in results.</param>
        /// <param name="userAgent">Optional User-Agent header value for outbound requests.</param>
        public HaloInfiniteClient(string spartanToken, string xuid = "", string clearanceToken = "", bool includeRawResponses = false, string userAgent = "")
        {
            this.SpartanToken = spartanToken;
            this.Xuid = xuid;
            this.ClearanceToken = clearanceToken;
            this.IncludeRawResponses = includeRawResponses;
            this.UserAgent = userAgent;

            this.InitializeModules();
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="HaloInfiniteClient"/> class with a custom HttpClient.
        /// </summary>
        /// <param name="httpClient">The HttpClient instance to use for API requests.</param>
        /// <param name="spartanToken">The Spartan token used for authentication.</param>
        /// <param name="xuid">Optional Xbox User ID.</param>
        /// <param name="clearanceToken">Optional clearance token for flighted content.</param>
        /// <param name="includeRawResponses">Whether to include raw API responses in results.</param>
        /// <param name="userAgent">Optional User-Agent header value for outbound requests.</param>
        public HaloInfiniteClient(
            HttpClient httpClient,
            string spartanToken,
            string xuid = "",
            string clearanceToken = "",
            bool includeRawResponses = false,
            string userAgent = "")
            : base(httpClient)
        {
            this.SpartanToken = spartanToken;
            this.Xuid = xuid;
            this.ClearanceToken = clearanceToken;
            this.IncludeRawResponses = includeRawResponses;
            this.UserAgent = userAgent;

            this.InitializeModules();
        }

        /// <summary>
        /// Gets the Academy module for bot customization and drill-related APIs.
        /// </summary>
        public AcademyModule Academy { get; private set; } = null!;

        /// <summary>
        /// Gets the Ban Processor module for ban-related APIs.
        /// </summary>
        public BanProcessorModule BanProcessor { get; private set; } = null!;

        /// <summary>
        /// Gets the Configuration module for endpoint discovery APIs.
        /// </summary>
        public ConfigurationModule Configuration { get; private set; } = null!;

        /// <summary>
        /// Gets the Economy module for player customization, stores, and inventory APIs.
        /// </summary>
        public EconomyModule Economy { get; private set; } = null!;

        /// <summary>
        /// Gets the Game CMS module for content management APIs including achievements, metadata, and files.
        /// </summary>
        public GameCmsModule GameCms { get; private set; } = null!;

        /// <summary>
        /// Gets the Lobby module for lobby and presence APIs.
        /// </summary>
        public LobbyModule Lobby { get; private set; } = null!;

        /// <summary>
        /// Gets the Settings module for clearance and flight configuration APIs.
        /// </summary>
        public SettingsModule Settings { get; private set; } = null!;

        /// <summary>
        /// Gets the Skill module for CSR and match skill APIs.
        /// </summary>
        public SkillModule Skill { get; private set; } = null!;

        /// <summary>
        /// Gets the Stats module for match history and service record APIs.
        /// </summary>
        public StatsModule Stats { get; private set; } = null!;

        /// <summary>
        /// Gets the Text Moderation module for moderation key APIs.
        /// </summary>
        public TextModerationModule TextModeration { get; private set; } = null!;

        /// <summary>
        /// Gets the UGC module for user-generated content authoring APIs.
        /// </summary>
        public UgcModule Ugc { get; private set; } = null!;

        /// <summary>
        /// Gets the UGC Discovery module for UGC search and discovery APIs.
        /// </summary>
        public UgcDiscoveryModule UgcDiscovery { get; private set; } = null!;

        private void InitializeModules()
        {
            this.Academy = new AcademyModule(this);
            this.BanProcessor = new BanProcessorModule(this);
            this.Configuration = new ConfigurationModule(this);
            this.Economy = new EconomyModule(this);
            this.GameCms = new GameCmsModule(this);
            this.Lobby = new LobbyModule(this);
            this.Settings = new SettingsModule(this);
            this.Skill = new SkillModule(this);
            this.Stats = new StatsModule(this);
            this.TextModeration = new TextModerationModule(this);
            this.Ugc = new UgcModule(this);
            this.UgcDiscovery = new UgcDiscoveryModule(this);
        }
    }
}
